effectful/handlers/pyro.py

```diff
@@ -330,29 +329,43 @@ def positional_distribution(
 ) -> tuple[TorchDistribution, Naming]:
     shape = d.shape()
     d = defterm(d)
-    dist_constr, args = d.args[0], d.args[1:]
 
-    if not (
-        d.op is call
-        and (
-            issubclass(dist_constr, TorchDistribution)
-            or issubclass(dist_constr, dist.torch_distribution.TorchDistributionMixin)
-        )
-    ):
+    if not isinstance(d, _DistributionTerm):
         raise NotImplementedError
 
     indices = sizesof(d).keys()
     naming = Naming.from_shape(indices, len(shape))
 
     def _to_positional(a):
         if isinstance(a, torch.Tensor):
+            a_indices = sizesof(a)
+            assert len(a_indices) == 0 or set(a_indices.keys()) == set(indices)
+            if len(a_indices) == 0:
+                return a
             return to_tensor(a, indices)
         elif isinstance(a, TorchDistribution):
             return positional_distribution(a)[0]
         else:
             return a
 
-    return dist_constr(*[_to_positional(a) for a in args], **d.kwargs), naming
+    new_d = d.op(
+        *[_to_positional(a) for a in d.args],
+        **{k: _to_positional(v) for (k, v) in d.kwargs.items()},
+    )
+
+    assert new_d.event_shape == d.event_shape
+    return new_d, naming
+
+
+@functools.cache
+def _register_distribution_op(
```

Comment:
Why was this necessary? Was it something to do with type inference and `call`?

Reply:
Yes, I think so, but I don't remember the precise problem. I also think the terms are simpler this way, and this change opens up the possibility of handling the distribution constructors.